### PR TITLE
init commit

### PR DESCRIPTION
**Pull request by Chengqian-Zhang**

### CODE REVIEW

File: tests/test_prep_run_vasp.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            self.assertEqual(out['backward_dir'] , Path(Path('task.%06d'%ii)/'my_backward'))
+            self.assertTrue(Path(out['backward_dir'] / 'my_log').is_file())
+            self.check_run_vasp_output(self.task_list_str[ii])
+
+@unittest.skipIf(skip_ut_with_dflow, skip_ut_with_dflow_reason)
+class TestPrepRunVasp(unittest.TestCase):
+    def setUp(self):
+        warnings.simplefilter('ignore', ResourceWarning)
+        self.ntasks = 3
+        self.confs = [Path(Path('confs')/'data.000'),Path(Path('confs')/'data.001')]
```
Comment (wanghan-iapcm) [inline tests/test_prep_run_vasp.py:138]: Do you really needs the conf files for the testing purpose?

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
- do we need examples on prep_vasp and run_vasp? a user is not supposed to use the steps at all. 
- please remove all the data files from the examples. Only provide the minimal set of files that are needed to run the examples.
- the examples are associated to the website of workflows.deepmodeling.com, which is not recommended. 
- python installation and distribution is missing
- the name convention of the modules and files are not good.

**Comment by Chengqian-Zhang** [comment]
> * do we need examples on prep_vasp and run_vasp? a user is not supposed to use the steps at all.
I have deleted the examples on prep_vasp and run_vasp.
> * please remove all the data files from the examples. Only provide the minimal set of files that are needed to run the examples.
I have updated the example of preprun_vasp. But It still has a "POTCAR" file. I don't know if this is appropriate.
> * the examples are associated to the website of workflows.deepmodeling.com, which is not recommended.
I have deleted the related infomation about the website of workflows.deepmodeling.com.
> * python installation and distribution is missing
I don't quite understand what "python installation and distribution" refers to.
> * the name convention of the modules and files are not good.
I have changed the structure and name of the directory. Please check again.

File: tests/test_prep_run_vasp.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        "run":{
+            "template_config" : {
+                "image" : default_image,
+            },  
+        },
+    }
+
+class TestMockedRunVasp(unittest.TestCase):
+    def setUp(self):
+        warnings.simplefilter('ignore', ResourceWarning)
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply tests/test_prep_run_vasp.py]: I have deleted the sentence.

File: tests/test_prep_run_vasp.py (hunk 1 above)
Comment (Chengqian-Zhang) [reply tests/test_prep_run_vasp.py]: I have deleted the conf files.

File: PrepRunFp/PrepRunFp.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        prep_config : dict,
+        run_config : dict,
+        upload_python_packages : Optional[List[os.PathLike]] = None,
+    ):
+        self._input_parameters = {
+            "config" : InputParameter(),
+            "inputs" : InputParameter(),
+            "optional_input" : InputParameter(),
+            "type_map" : InputParameter(),
+            "backward_list" : InputParameter(),
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply PrepRunFp/PrepRunFp.py]: I change them to:
"optional_input" : InputParameter(type=dict , value={})
"inputs","type_map" and "backward_list" are mandatory.

File: PrepRunFp/PrepRunFp.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        prep_op : OP,
+        run_op : OP,
+        prep_config : dict,
+        run_config : dict,
+        upload_python_packages : Optional[List[os.PathLike]] = None,
+    ):
+        self._input_parameters = {
+            "config" : InputParameter(),
+            "inputs" : InputParameter(),
+            "optional_input" : InputParameter(),
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply PrepRunFp/PrepRunFp.py]: I change them to:
"config" : InputParameter(type=dict , value={})
"optional_input" : InputParameter(type=dict , value={})
"inputs" is mandatory.
Optional parameters have been placed after compulsory parameters.

File: PrepRunFp/PrepRunFp.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            "type_map" : InputParameter(),
+            "backward_list" : InputParameter(),
+            "log_name" : InputParameter(type=str , value="log"),
+            "backward_dir_name" : InputParameter(type=str , value="backward_dir"),
+        }
+        self._input_artifacts = {
+            "confs" : InputArtifact(),
+            "optional_artifact" : InputArtifact(),
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply PrepRunFp/PrepRunFp.py]: I change them to:
"confs" : InputArtifact(),
"optional_artifact" : InputArtifact(optional=True),

File: tests/test_prep_run_vasp.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        for ii in range(self.ntask):
+            ip = OPIO({
+                'task_name' : self.task_list_str[ii],
+                'task_path' : self.task_list[ii],
+                'log_name' : "my_log",
+                'backward_dir_name' : "my_backward",
+                "config" : {},
+                "optional_artifact" : {},
+                "backward_list" : ['POSCAR','POTCAR'],
+                "optional_input" : {},
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply tests/test_prep_run_vasp.py:147]: I have added tests for the case when the optional parameters/artifacts are not provided.Please check.

File: LICENSE [hunk 7]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (njzjz) [inline LICENSE:1]: MIT is not compatible with LGPL

File: fpop/run_fp.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    dargs,
+    Argument,
+    Variant,
+    ArgumentEncoder,
+)
+
+@contextmanager
+def set_directory(path: Path):
+    '''Sets the current working path within the context.
+    Parameters
```
Comment (njzjz) [inline fpop/run_fp.py:45]: There should be a blank line be fore each section

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+setuptools.setup(
+    name="fpoptest",
```
Comment (njzjz) [inline setup.py:7]: Do we want to use this name?

File: fpop/preprun_fp.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+class PrepRunFp(Steps):
+    def __init__(
+        self,
+        name : str,
+        prep_op : OP,
+        run_op : OP,
+        prep_image : str,
+        run_image : str,
+        prep_config : Optional[dict] = None,
+        run_config : Optional[dict] = None,
```
Comment (wanghan-iapcm) [inline fpop/preprun_fp.py:42]: It seems that the input config should not be one simple dict. 
The parameters passed to template, slice and the step may be provided separately.

File: fpop/preprun_fp.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        upload_python_packages : Optional[List[os.PathLike]] = None,
+    ):
+        self._input_parameters = {
+            "inputs" : InputParameter(),
+            "type_map" : InputParameter(),
+            "backward_list" : InputParameter(),
+            "config" : InputParameter(type=dict , value={}),
+            "optional_input" : InputParameter(type=dict , value={}),
```
Comment (wanghan-iapcm) [inline fpop/preprun_fp.py:50]: Please provide doc explaining what are the acceptable keys in the optional input.

File: fpop/preprun_fp.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            "log_name" : InputParameter(type=str , value="log"),
+            "backward_dir_name" : InputParameter(type=str , value="backward_dir"),
+        }
+        self._input_artifacts = {
+            "confs" : InputArtifact(),
+            "optional_artifact" : InputArtifact(optional=True),
```
Comment (wanghan-iapcm) [inline fpop/preprun_fp.py:56]: Please provide doc explaining the `optional_artifact`

File: fpop/preprun_fp.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        run_image : str,
+        prep_config : Optional[dict] = None,
+        run_config : Optional[dict] = None,
+        upload_python_packages : Optional[List[os.PathLike]] = None,
+    ):
+        self._input_parameters = {
+            "inputs" : InputParameter(),
+            "type_map" : InputParameter(),
+            "backward_list" : InputParameter(),
+            "config" : InputParameter(type=dict , value={}),
```
Comment (wanghan-iapcm) [inline fpop/preprun_fp.py:49]: we need one config for prep and one for run.

The name of the `config` may be confused with the `config` of `dflow.Step`. We may rename it as `image_config`

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: LICENSE (hunk 7 above)
Comment (Chengqian-Zhang) [reply LICENSE]: I have changed LICENSE to LGPL.

File: fpop/run_fp.py (hunk 8 above)
Comment (Chengqian-Zhang) [reply fpop/run_fp.py:45]: I have added the blank lines.

File: setup.py (hunk 9 above)
Comment (Chengqian-Zhang) [reply setup.py:7]: I change the name to fpop.But I have not submitted to pypi.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,2 @@
+# FP_OPs
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply README.md]: I have updated README.

File: fpop/preprun_fp.py (hunk 13 above)
Comment (Chengqian-Zhang) [reply fpop/preprun_fp.py]: I split this "config" to "prep_image_config" and "run_image_config".For example, "command" is a keyward in "run_image_config"

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        pip install -e .[test]
+        pip install mock coverage pytest
+    - name: Test
+      run: |
+        export DFLOW_HOST=${{ secrets.DFLOW_HOST }}
+        export DFLOW_K8S_API_SERVER=${{ secrets.DFLOW_K8S_API_SERVER }}
+        export DFLOW_S3_REPO_KEY=${{ secrets.DFLOW_S3_REPO_KEY }}
+        export DFLOW_S3_STORAGE_CLIENT=${{ secrets.DFLOW_S3_STORAGE_CLIENT }}
+        export BOHRIUM_USERNAME=${{ secrets.BOHRIUM_USERNAME }}
+        export BOHRIUM_PASSWORD=${{ secrets.BOHRIUM_PASSWORD }}
```
Comment (AnguseZhang) [inline .github/workflows/test.yml:31]: Delete these if you do not need.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,2 @@
+# fpop
+`fp` stands for first principle calculation and `op` stands for operators. The abbreviation `fpop` stands for operators related to first principle calculation.
```
Comment (AnguseZhang) [inline README.md:2]: first-principles

Comment (AnguseZhang) [reply README.md:2]: Suggest add some basic background here, such as based on `dflow` (attached link). Furnish the words.

File: fpop/prep_fp.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+    >>> with set_directory("some_path"):
+    ...    do_something()
+    '''
+    cwd = Path().absolute()
+    path.mkdir(exist_ok=True, parents=True)
+    try:
+        os.chdir(path)
+        yield
+    finally:
+        os.chdir(cwd)
```
Comment (AnguseZhang) [inline fpop/prep_fp.py:47]: Use the tools `set_directory` in `dflow`

**Review by AnguseZhang — CHANGES_REQUESTED** [review]

File: fpop/prep_fp.py (hunk 17 above)
Comment (Chengqian-Zhang) [reply fpop/prep_fp.py]: I have done it.

File: .github/workflows/test.yml (hunk 15 above)
Comment (Chengqian-Zhang) [reply .github/workflows/test.yml]: The bug of dflow is fixed.

File: setup.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+setuptools.setup(
+    name="fpop",
+    version="0.0.3",
+    author="Chengqian Zhang",
+    author_email="2043899742@qq.com",
+    description="operators about first principle caculations",
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    url="https://github.com/Chengqian-Zhang/FP_OPs",
```
Comment (AnguseZhang) [inline setup.py:14]: Update url and necessary information

**Review by wanghan-iapcm — APPROVED** [review]

**Comment by AnguseZhang** [comment]
@Chengqian-Zhang Please solve the minor changes I requested. Then I will merge this PR.

File: README.md [hunk 19]
```diff
@@ -0,0 +1,2 @@
+# fpop
+`fp` stands for first-principles calculation and `op` stands for operators. The abbreviation `fpop` stands for operators related to first-principles calculation.This project is base on [dflow](https://github.com/deepmodeling/dflow) which is a Python framework for constructing scientific computing workflows (e.g. concurrent learning workflows) employing Argo Workflows as the workflow engine.
```
[the remark this answers is not in the document]
Comment (Chengqian-Zhang) [reply README.md]: I have fixed it.

File: setup.py (hunk 18 above)
Comment (Chengqian-Zhang) [reply setup.py]: I have updated url.

**Pull request merged**